pages/design_documentation/cleaner.md

```diff
@@ -5,12 +5,147 @@ toc: false
+If all the conditions are met, a flush list of up to **Flush Max Buffers** 
+is built by removing the cache lines from the back of the LRU list if their 
+**staleness time** has expired.
+
+This list is then provided to the OCF cleaner by issuing *ocf_cleaner_fire()*. 
+OCF cleaner then issues an asynchrounous flushing request, eventually writing 
+the dirty data to the core device.
+
+
 ### ACP - Aggressive Cleaning Policy
-### NOP - No Operation (effectively disables background cleaning)
+Aggressive Cleaning Policy takes a rather different approach than ALRU. It organizes
+the core device into 100MiB chunks, maintaining a list of them. The chunks are put 
+into "buckets" of varying dirty data percentage. There are 11 buckets in the ACP, 
+holding data chunks being 0% dirty, up to 100% dirty, with a 10% granularity. ACP
+cleans the dirtiest chunks first, as described below.
+
+
+#### Configuration
+ACP uses these config parameters to tune the behavior of the cleaner.
+
+* **Wake Up Time** - cleaning idle sleep wake up time. (default: 10s)
+
+* **Flush Max Buffers** - maximum amount of dirty cache lines to flush in one cleaning 
+operation (default: 128 lines)
+
+
+#### Operation
+ACP creates the chunks lists per each core as a part of core adding process.
+A chunk represents a linear, 100MiB part of the core device address space.
+This creates an ACP-specific mapping of cache lines to the chunks. These chunks
+are then put into a proper buckets, depending on the percentage of dirty cache 
+lines in a chunk.
+
+When a cache line is written and dirty data is created, the dirty lines counter 
+in the chunk the cache line belongs to is incremented. Since the chunks are placed 
+in the buckets according to the dirty lines percentage in a chunk, a chunk may be
+moved to another bucket once the percentage threshold is met.
+
+After **Wake Up Time** time has passed, background cleaning needs to be called by 
+the user by issuing *ocf_cleaner_run()*. In the ACP, the cleaning logic selects 
+the chunk to clean starting form the buckets with the higest dirty data percentage.
+As soon as there is a non-empty chunk list within a bucket, first chunk from that 
+list is selected.
+
+The list of up to **Flush Max Buffers** of cache lines is provided to the OCF
+cleaner. The *ocf_cleaner_fire()* is called and the OCF cleaner issues an 
+asynchrounous flushing request, eventually writing the dirty data to the core 
+device.
+
+After a successful flush, the *ocf_cleaner_run()* callback will be called with the 
+**Wake Up Time** time interval parameter, and the selection of a chunk to clean 
+will commence again. If the chunk that was previously cleaned still has some dirty
+data, it will be selected again, the cleaning constraints will be evaluated again 
+and the next portion of up to **Flush Max Buffers** of cache lines of this chunk 
+will be flushed.
+
+
+### NOP - No Operation
+No operation is performed by the background cleaner task, no logic is called to 
```

Comment:
For consistency with above paragraphs I would write:
"No Operation Policy (NOP) is equivalent to no steps being performed in the background by the cleaner task."